doc/kuksa_analysis/kuksa_design_topics.md

```diff
@@ -0,0 +1,744 @@
+
+#### Bidirectional stream
+
+A bidirectional stream between Provider <-> databroker
+- Provides databroker with a way to associate information sent from the provider
+  (e.g. capabilities, which actuators it provides, which sensors etc) with the stream
+  it uses to control it.
+
+#### Actuators
+VSS defines three types of signals:
+* Attribute
+* Sensor
+* Actuator
+
+An actuator acts as both something you can actuate and something providing values (a sensor).
+It's not even necessarily the same component providing the implementation of these separate concerns. With this in mind, a provider providing an VSS actuator would in this design provide
+an `Actuator(path)` to the server in order to receive `ActuateCommand`s, and provide
+`Sensor(path)` and send `SensorData` when they are providing sensor data.
+
+The alternative would be to duplicate everything in `Sensor` for `Actuator`s.
+
+
+## Overview
+The stream in each direction would consist of different "messages" implemented with `oneof {...}`.
+
+In the direction of *Provider -> Server*, at least three types of "messages" would flow<sup>(2)</sup>:
+* `SensorData` containing data streamed by the provider.
+* `Sensor` containing sensor information from the provider.
+* `Actuator` containing actuator information from the provider.
+* `Attribute` containing attribute including the value. <sup>(1)</sup>
+
+In the direction of *Server -> Provider*, at least three types of "messages" would flow:
+* `ActuateCommand`, tells actuator to actuate (triggered by someone setting actuator targets).
+* `SensorCommand`, controls behaviour of a sensor, i.e.
+   "start", "stop", "try to use this update frequency" etc..
+* `Error`, an error occurred. One type at the moment. It would probably make sense to
+   split it into errors that can occur at any time & errors that are directly caused by
+   things the provider provided.
+
+<sup>(1)</sup> It would probably make sense to introduce a separate `Error` in the direction of
+    *Provider -> Server*. Currently, the only errors in that direction is `ReadError` as part of the
+    sensor data.
+
+<sup>(2)</sup> It's possible that it makes mores sense to provide a separate unary RPC call for setting attributes,
+    since attributes (probably) don't update frequently and (probably) wont need availability status etc..
+
+```proto
+service VAL {
+  ...
+  rpc Provide(stream ProviderRequest) returns (stream ProviderResponse);
+}
+
+message ProviderRequest {
+  oneof provide {
+    Sensor sensor = 1;
+    Actuator actuator = 2;
+    Attribute attribute = 3;
+    SensorData sensor_data = 4;
+  }
+}
+
+message ProviderResponse {
+  oneof command {
+    ActuateCommand actuator_command = 1;
+    SensorCommand sensor_command = 2;
+    Error error = 3;
+  }
+}
+
+...
+
+```
+
+The full `.proto` specification for this proposal can be found [here](https://github.com/boschglobal/kuksa.val/blob/feature/provider_api/kuksa_databroker/proto/sdv/databroker/v1/provider.proto).
```

Comment:
This link is dead and caused CI error. In general search for all `boschglobal` and `kuksa.val`, none of them shall be used.